backend/server.js

```diff
@@ -274,11 +359,70 @@ app.post("/users/:id/change-password", (req, res) => {
     .then(() => {
       res.status(201).json({ message: "Käyttäjän salasana tallennettu" });
     })
-    .catch((error) => {
+    .catch(() => {
       res.status(500).json({ message: "Virhe!" });
     });
 });
 
+async function validateToken(req, res, next) {
+  const idToken = req.headers.authorization?.split("Bearer ")[1];
+  if (!idToken) {
+    return res.status(401).json({ message: "Unauthorized" });
+  }
+
+  try {
+    const decodedToken = await getAuth().verifyIdToken(idToken);
+    req.user = decodedToken;
+    next();
+  } catch (error) {
+    return res.status(401).json({ message: "Invalid token" });
+  }
+}
+app.post("/users/:id/permissions", validateToken, async (req, res) => {
+  if (req.user.admin !== true) {
+    return res.status(401).json({ message: "Unauthorized" });
+  }
+
+  const uid = req.params.id;
+  const userRecord = await getAuth().getUser(req.user.uid);
```

Comment:
You're getting the user who made the request rather than the user whose permissions are modified. Check what uid should be used.